expat/lib/xmlparse.c

```diff
@@ -6295,6 +6295,30 @@ poolStoreString(STRING_POOL *pool, const ENCODING *enc,
   return pool->start;
 }
 
+static size_t
+poolBytesToAllocateFor(int blockSize)
+{
+  /* Unprotected math would be:
+  ** return offsetof(BLOCK, s) + blockSize * sizeof(XML_Char);
+  **
+  ** Detect overflow, avoiding _signed_ overflow undefined behavior
+  ** For a + b * c we check b * c in isolation first, so that addition of a
+  ** on top has no chance of making us accept a small non-negative number
+  */
+  const int stretchedBlockSize = (int)((unsigned)blockSize * sizeof(XML_Char));
+  if (stretchedBlockSize < 0)
+    return 0;
+
+  {
+    const int bytesToAllocate = (int)(
+        offsetof(BLOCK, s) + (unsigned)stretchedBlockSize);
+    if (bytesToAllocate < 0)
```

Comment:
Since the two lines suggested above leave you with a positive `stretchedBlockSize` (assuming a positive `blockSize` argument), this check suffices.